sycl/test-e2e/Matrix/joint_matrix_colA_rowB_colC_impl.hpp

```diff
@@ -0,0 +1,130 @@
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental::matrix;
+using bfloat16 = sycl::ext::oneapi::bfloat16;
+
+#define TM 8
+#define TN SG_SZ
+#define TK 16
+
+#define BF16_EPSILON 0.00781250
+
+float make_fp32(bfloat16 x) {
+  unsigned int y = *((int *)&x);
+  y = y << 16;
+  float *res = reinterpret_cast<float *>(&y);
+  return *res;
+}
+
+template <typename T1, typename T2, size_t NUM_ROWS_A, size_t NUM_COLS_A,
+          size_t NUM_ROWS_B, size_t NUM_COLS_B, size_t NUM_ROWS_C,
+          size_t NUM_COLS_C>
+void matrix_multiply(T1 *C, T2 *A, T2 *B, queue q) {
+  size_t M = NUM_ROWS_C;
+  size_t N = NUM_COLS_C;
+  size_t K = NUM_COLS_A;
+
+  assert(NUM_ROWS_C == NUM_ROWS_A && NUM_COLS_A == NUM_ROWS_B);
+  size_t NDRangeM = M / TM;
+  size_t NDRangeN = N / TN;
+
+  auto pA = multi_ptr<T2, sycl::access::address_space::global_space>(A);
+  auto pB = multi_ptr<T2, sycl::access::address_space::global_space>(B);
+  auto pC = multi_ptr<T1, sycl::access::address_space::global_space>(C);
+
+  q.submit([&](handler &cgh) {
+     cgh.parallel_for<class imatrix>(
+         nd_range<2>({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ}),
+         [=](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]]
+
+         {
+           // The submatrix API has to be accessed by all the workitems in a
+           // subgroup these functions will be called once by the subgroup no
+           // code divergence between the workitems
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<sub_group, bfloat16, use::a, TM, TK, layout::col_major>
+               sub_a;
+           joint_matrix<sub_group, bfloat16, use::b, TK, TN, layout::row_major>
+               sub_b;
+           joint_matrix<sub_group, float, use::accumulator, TM, TN> sub_c;
+           joint_matrix_fill(sg, sub_c, 1);
+           for (int k = 0; k < K; k += TK) {
+             joint_matrix_load(sg, sub_a, pA + (sg_startx * TM) * K + k, K);
+             joint_matrix_load(
+                 sg, sub_b, pB + (k) * (N) + sg_starty / SG_SZ * TN * 2, N * 2);
+             sub_c = joint_matrix_mad(sg, sub_a, sub_b, sub_c);
+           }
+           joint_matrix_store(
+               sg, sub_c, pC + (sg_startx * TM) * N + sg_starty / SG_SZ * TN, N,
+               layout::col_major);
+         }); // parallel for
+   }).wait();
+}
+
+void matrix_multiply_ref(bfloat16 *A, bfloat16 *B, float *C, int MATRIX_M,
+                         int MATRIX_N, int MATRIX_K) {
+  for (unsigned int i = 0; i < MATRIX_M; i++) {
+    for (unsigned int k = 0; k < MATRIX_K; k++) {
+      for (unsigned int j = 0; j < MATRIX_N; j++) {
+        C[j * MATRIX_N + i] +=
```

Comment:
This whole function should be in common.hpp to avoid future errors.
```suggestion
        C[j * MATRIX_M + i] +=
```

Reply:
This one is different as it stores C in column major. So in common.hpp, we would have two functions:
matrix_multiply_ref
and
matrix_multiply_ref_transposed_c

Reply:
Oh, OK, good to know!